Runtime/Scripts/WebRTC.cs

```diff
@@ -804,36 +918,56 @@ internal static class NativeMethods
         public static extern void PeerConnectionRestartIce(IntPtr ptr);
         [DllImport(WebRTC.Lib)]
         public static extern RTCErrorType PeerConnectionSetConfiguration(IntPtr ptr, [MarshalAs(UnmanagedType.LPStr, SizeConst = 256)] string conf);
+#if !UNITY_WEBGL
         [DllImport(WebRTC.Lib)]
         public static extern IntPtr ContextCreateDataChannel(IntPtr ptr, IntPtr ptrPeer, [MarshalAs(UnmanagedType.LPStr, SizeConst = 256)] string label, ref RTCDataChannelInitInternal options);
+#else
+        [DllImport(WebRTC.Lib)]
+        public static extern IntPtr ContextCreateDataChannel(IntPtr ptr, IntPtr ptrPeer, [MarshalAs(UnmanagedType.LPStr, SizeConst = 256)] string label, [MarshalAs(UnmanagedType.LPStr, SizeConst = 256)] string optionsJson);
```

Comment:
I am not familiar the interoperability for WebGL and Unity, 😛 
That's why sorry to ask such a basic question.

- Can not we use `ref`, `out` keyword to communicate WebGL plugin ?
- I am sure that we need to use JSON to pass the struct data to WebGL plugin from Unity. Is there any other ways to pass them?

Reply:
You can use ref and out, but since they are passed to JS with pointers, you can basically only pass values as numbers, strings, or arrays of numbers.
Also, the method of passing the structure can be done by creating a Serializer / Deserializer by yourself other than JSON, but the performance of the JSON API built in JS is better than the poor implementation of Serializer / Deserializer by yourself. 

(日本語)
一応、ref や out は使えますが、JSにはポインターで渡されるので基本的に数値か文字列、もしくは数値配列でしか値を渡せません。
構造体を渡す方法は、バイト配列を渡すことが可能なので、自分でSerializer/Deserializerを作成することでJSON以外でも可能にはなりますが、下手に自分でSerializer/Deserializerを実装するよりもJS組み込みのJSON APIのほうがパフォーマンスがいいです。

Reply:
As gtk2k said, we can only send basic types to jslib.  [Here](https://docs.unity3d.com/Manual/webgl-interactingwithbrowserscripting.html) it states the supported types (strings, different numbers in different HEAP arrays and a webgltextures map).
So unfortunately, sending json objects (strings) back and forth it the only way to go (at least to my knowledge) for these type of connectivity.

I did found on the emscripten docs that it's possible to send an array of pointers, which could be useful for the larger webrtc stats and/or rtc parameter classes and let the strings stay in the javascript heap and acces them from unity with pointers.

Reply:
@gtk2k, @nvanofwegen
Talking to you is very helpful to me. Thanks.

I would like to reduce conditional branches using **UNITY_WEBGL** in `WebRTC.cs` and `Context.cs` to keep code easy to maintenance. Do you have any ideas to achieve it? There is no problem for me to rewrite trunk source code for it.

Reply:
Yeah, the conditional branches are a pain... But I'm not sure if there is a good solution that is worth the struggle, just to have the code a little bit more maintainable. For now I'd say it is the best to have it better structured in WebRTC.cs, so the overview is a bit more clear, which makes it also easier the maintain. Right now the NativeMethods have the UNITY_WEBGL all over it, I agree this is a bit confusing.

What I found/think:
As unity and javascript have their own memory heap, it's afaik not possible to simply 'share' the objects between unity and javascript. Only the available heap arrays are an option (numbers, bytes, textures).  Strings are not shared as far as i know, just pointers to read the data.
gtk2k said that you could write a serializer/deserializer yourself, instead of json, using byte heap/arrays. Regarding that, also found [this thread](https://forum.unity.com/threads/source-code-c-javascript-interoperability-library.968487/) where a serializer/deserializer is used instead of json.

But, as this is not really my playing field, I have no idea if this requires a lot of c++ changes or not. It also looks like the functionality that is needed for the RTCParameter/RTCStats adoption, so maybe in the long run this has to be done anyway. If that is the case, I guess also implementing the ref interop functionality is pretty much the same?

Some other interesting thread I found, but it is maybe a bit outdated:
https://forum.unity.com/threads/trying-to-share-float-array-between-c-and-webgl-browser-javascript.394861/

[Here](https://forum.unity.com/posts/2579401) the threads starts about strings being immutable and [here](https://forum.unity.com/posts/2586907) it's even stated that it's simple not possible. But, as this thread is from 2016, so maybe now it's possible?

Reply:
Thanks.

> Here the thread starts about strings being immutable and here it's even stated that it's simple not possible. But, as this thread is from 2016, so maybe now it's possible?

I am not sure that the currently status but it would not be supported yet.